### PR TITLE
Tags train

### PR DESCRIPTION
**Pull request by LingLing10**
### Description of change
Add a new pipeline for tags classifier training. So there will be two pipelines, one is for training, another one is for prediction (which is already live).

### Checklist

* [ ] Have tests been added to cover any changes?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: dataflow/dags/tags_classifier_train_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            ("precisions", sa.Column("precisions", sa.Numeric)),
+            ("recalls", sa.Column("recalls", sa.Numeric)),
+            ("f1", sa.Column("f1", sa.Numeric)),
+            ("accuracy", sa.Column("accuracy", sa.Numeric)),
+            ("auc", sa.Column("auc", sa.Numeric)),
+        ],
+    )
+
+    # controller_pipeline = InteractionsDatasetPipeline ## Not in use
+    # dependencies = [InteractionsDatasetPipeline] ## the pipeline that uploads training data from BIU
```
Comment (abbas123456) [inline dataflow/dags/tags_classifier_train_pipeline.py:30]: Can we delete this commented out code?

File: dataflow/dags/tags_classifier_train_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    def get_fetch_operator(self) -> PythonOperator:
+        return PythonOperator(
+            task_id='train-model',
+            python_callable=model_training_with_labelled_data,
+            queue='tensorflow',
+            provide_context=True,
+            # op_args=[self.target_db,  self.table_config.table_name],
```
Comment (abbas123456) [inline dataflow/dags/tags_classifier_train_pipeline.py:38]: Can we delete this commented out code?

File: dataflow/operators/tags_classifier/train/train.py [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+
+    tokenizer = Tokenizer(
+        num_words=MAX_NB_WORDS,
+        filters='!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n',
+        lower=True,
+    )
+
+    tokenizer.fit_on_texts(df['sentence'].values)
+    word_index = tokenizer.word_index
+    logger.info(f'Found %s unique tokens.{len(word_index)}')
```
Comment (abbas123456) [inline dataflow/operators/tags_classifier/train/train.py:74]: This needs to be
`logger.info('Found %s unique tokens', len(word_index))`

File: dataflow/operators/tags_classifier/train/train.py [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+    from tensorflow.keras.models import Sequential
+    from tensorflow.keras import layers
+    from tensorflow.keras.layers import Embedding
+
+    model = Sequential()
+    # model.add(layers.Embedding(input_dim=vocab_size,
+    #                            output_dim=embedding_dim,
+    #                            input_length=maxlen))
+    # model.add(Embedding(MAX_NB_WORDS, EMBEDDING_DIM, input_length=X.shape[1],trainable=True,mask_zero=True, weights=[embedding_matrix]))
+    # model.add(Embedding(MAX_NB_WORDS, EMBEDDING_DIM, input_length=X.shape[1],trainable=True, weights=[embedding_matrix]))
```
Comment (abbas123456) [inline dataflow/operators/tags_classifier/train/train.py:133]: Again can we remove this commented out code

File: dataflow/operators/tags_classifier/train/train.py [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+        logger.info("step 2: train model")
+        build_models_pipeline(df, train_data_date)
+
+        logger.info("step 3: save model")
+        save_model(train_data_date)
+
+        logger.info("step 4: write model performance to S3")
+        write_model_performance(table_name, train_data_date, **context)
+
+    pass
```
Comment (abbas123456) [inline dataflow/operators/tags_classifier/train/train.py:186]: This isn't needed

File: dataflow/operators/tags_classifier/train/train.py [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+            logger.info(f'check class_weight:  {class_weight}')
+
+            m = cnn()
+            m = train_model(m, X_train, Y_train_tag, X_val, Y_val_tag, class_weight)
+            m.save(model_path + '_'.join(j.split(' ')))
+
+            data_to_evaluate = X_test
+            tag_to_evaluate = Y_test_tag
+            sent_to_evaluate = sent_test
+            # data_to_evaluate = X_val
```
Comment (abbas123456) [inline dataflow/operators/tags_classifier/train/train.py:286]: Again can we remove this commented out code

File: dataflow/operators/tags_classifier/train/train.py [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+    tag_auc = dict()
+    tag_size = dict()
+
+    Y_test_predict = np.zeros(Y_test.shape)
+    Y_test_predict_prob = np.zeros(Y_test.shape)
+
+    logger.info(f"build models for tags:{tags}")
+
+    for j in tags:
+        if 1 == 1:
```
Comment (abbas123456) [inline dataflow/operators/tags_classifier/train/train.py:258]: This isn't needed

File: dataflow/operators/tags_classifier/train/train.py [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+
+    # tn_ind = np.concatenate((test_predictions_prob_tag1<probability_threshold)&(Y_test[:,tag_i]==0))
+    tn_sen = sent_test[
+        (np.concatenate(test_predictions_prob_tag1) <= probability_threshold)
+        & (Y_test[:, tag_i] == 0)
+    ]
+    tn_p = test_predictions_prob_tag1[
+        np.concatenate((test_predictions_prob_tag1 < probability_threshold))
+        & (Y_test[:, tag_i] == 0)
+    ]
```
Comment (abbas123456) [inline dataflow/operators/tags_classifier/train/train.py:383]: `fp_sen, tp_sen, fn_sen, tn_sen, fp_p, tp_p, fn_p, tn_p` all seem to be set to the same thing in this function. Are they meant to be the same?

**Comment by abbas123456** [comment]
Overall it looks good, theres a lot of commented out code that needs to be removed and a couple of points I have mentioned.

File: dataflow/dags/tags_classifier_train_pipeline.py (hunk 1 above)
Comment (LingLing10) [reply dataflow/dags/tags_classifier_train_pipeline.py:30]: done

File: dataflow/dags/tags_classifier_train_pipeline.py (hunk 2 above)
Comment (LingLing10) [reply dataflow/dags/tags_classifier_train_pipeline.py:38]: done

File: dataflow/operators/tags_classifier/train/train.py (hunk 3 above)
Comment (LingLing10) [reply dataflow/operators/tags_classifier/train/train.py:74]: Thank you for pointing this out. change to `logger.info(f'Found {len(word_index)} unique tokens')`

File: dataflow/operators/tags_classifier/train/train.py (hunk 4 above)
Comment (LingLing10) [reply dataflow/operators/tags_classifier/train/train.py:133]: done

File: dataflow/operators/tags_classifier/train/train.py (hunk 5 above)
Comment (LingLing10) [reply dataflow/operators/tags_classifier/train/train.py:186]: Remove the 'pass'

File: dataflow/operators/tags_classifier/train/train.py (hunk 6 above)
Comment (LingLing10) [reply dataflow/operators/tags_classifier/train/train.py:286]: done.

File: dataflow/operators/tags_classifier/train/train.py (hunk 7 above)
Comment (LingLing10) [reply dataflow/operators/tags_classifier/train/train.py:258]: remove the 'if 1==1'

File: dataflow/operators/tags_classifier/train/train.py (hunk 8 above)
Comment (LingLing10) [reply dataflow/operators/tags_classifier/train/train.py:383]: This function is for sanity checking, not used in production but for development. The function is kept here so if we want to check the results we can use it. fp_sen, tn_sen, fp_p, tn_p, tp_sen,fn_sen, tp_p, fn_p are defined differently based on two boolean values.

**Comment by LingLing10** [comment]
Update code based on the review comments and have pushed the update.

**Comment by LingLing10** [comment]
Hi, I have changed to code a bit to make the pipeline automatically pick up the latest data for training and latest model version for prediction. Could you please review it? And if things are OK, can we merge it to master? Thank you!

**Review by abbas123456 — APPROVED** [review]

**Pull request merged**